### ISSUE TITLE
How to set Restart=always for a socket unit?

### ISSUE THREAD

**Issue by lahwaacz** [systemd/systemd#11553]
Even socket units may fail, e.g. due to an out-of-memory situation, which may trigger errors like
```
systemd[1]: sshd.socket: Failed to fork off accept stub process: Cannot allocate memory
```
(The unit comes from Arch Linux [openssh package](https://git.archlinux.org/svntogit/packages.git/tree/trunk/sshd.socket?h=packages/openssh).)

Since there is no "nice way to fix the root cause", setting `Restart=always` for the unit is needed to prevent losing access to headless hosts with difficult physical access, but that clause is valid only for services. Is there another way to have socket units always restart on failure?

**Comment by mailinglists35** [comment]
what is the workaround until a native solution is provided? create a service to check the socket unit?

**Comment by lahwaacz** [comment]
@travier That is very different mechanism than restarting. `TriggerLimitIntervalSec=0` prevents the socket from entering failed state only due to hitting a trigger limit, but not due to e.g. out-of-memory conditions.

**Comment by travier** [comment]
Good point. This only works for the DoS case.

**Comment by YHNdnzj** [comment]
Try to change the dependency type to `Upholds=` (UpheldBy=)? While it's not customizable based on specific error, the use case of `Restart=always` greatly.

**Comment by lahwaacz** [comment]
If it works, the "fake" unit can be even `sockets.target`.

**Comment by YHNdnzj** [comment]
> Looks like https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Upholds= could indeed be workaround but it's not great as it require another "fake" unit.

Why is something like that required? You can specify `UpheldBy=` and enable the socket as usual, no?

**Comment by YHNdnzj** [comment]
> OK, so something like below could potentially do it (needs testing) for the `sshd.socket` unit (or any other socket unit):
> ```
> [Unit]
> UpheldBy=sockets.target
> ```

Just like `WantedBy=`, `UpheldBy=` is an `[Install]` setting. You should simply replace WantedBy with UpheldBy and `reenable` the socket.

**Comment by travier** [comment]
Good point, updated.

**Comment by travier** [comment]
For reference, https://github.com/systemd/systemd/pull/29159 should help with this.

**Comment by dustinlagoy** [comment]
I had an issue with a socket failing on boot, due to a network resource not yet being available, that could be worked around easily if this issue were resolved. I tried adding the socket to the `Upholds=` setting of `sockets.target`. This worked but doesn't solve the issue because the socket would restart quickly and hit a restart limit before the network resource came online. Setting the `StartLimitIntervalSec` to 0 for the socket doesn't help much because there is apparently a hardcoded check for 16 restarts within a 10 second interval (this is `Unit.auto_start_stop_ratelimit` currently set at https://github.com/systemd/systemd/blob/f660c7fa56b247c278fdb2ebcfea37912f249524/src/core/unit.c#L137).

For what it is worth the error due to the rate limit is:
```
Unit needs to be started because active unit sockets.target upholds it, but not starting since we tried this too often recently.
```

I worked around it with another service to monitor and restart the socket but it would be nice if systemd would handle this.

@travier I'm not sure #29159 would help here.

**Comment by travier** [comment]
@dustinlagoy This is not the same issue. If your socket depends on a specific network resource being up then you should order it after that, not bypass/workaround the retry logic in systemd.

**Comment by dustinlagoy** [comment]
Yes you are correct in my case that is a better solution. Perhaps for other cases there may be some need to work around the hardcoded rate limit though I don't know what they would be.
